File: D3/basic.py
from browser import *
from workbench import *
import math

d3 = window.d3

width = max(960, window.innerWidth)
height = max(500, window.innerHeight)
x1 = width / 2
y1 = height / 2
x0 = x1
y0 = y1
i = 0
r = 200
tau = math.pi * 2

canvas = d3.select("body").append("canvas")

canvas.attr("width", width).attr("height", height)

d3.select("canvas").style("background-color", "#222")

def move(unused1, unused2, unused3):
    global x1, y1
    mouse = d3.mouse(canvas.node())
    x1 = mouse[0]
    y1 = mouse[1]
    d3.event.preventDefault()

canvas.on("mousemove", move)

workbench = Workbench2D(canvas.node())

context = canvas.node().getContext("2d")
context.globalCompositeOperation = "lighter"
context.lineWidth = 2

def setUp():
    workbench.setUp()

def tick(t):
    global i, x0, y0
    i += 1
    context.clearRect(0,0,width,height)
    z = d3.hsl(i % 360, 1, 0.5).rgb()
    c = "rgba(" + str(int(z.r)) + "," + str(int(z.g)) + "," + str(int(z.b)) + ","
    x0 += (x1 - x0) * .1
    y0 += (y1 - y0) * .1
    x = x0
    y = y0
    def tweeny(unused1, unused2):
        # This side effect is necessary to make things work!
        a = x
        def circle(t):
            s = c + str(1-t) + ")"
            context.strokeStyle = s
            context.beginPath()
            context.arc(x, y, r * t, 0, tau)
            context.stroke()
        return circle
    d3.select({}).transition().duration(2000).ease(math.sqrt).tween("circle", tweeny)

def terminate(t):
    return t > 60
    
def tearDown(e):
    workbench.tearDown()
    if e:
        print "Ending with exception: " + str(e)

war = WindowAnimationRunner(tick, terminate, setUp, tearDown, window)
war.start()

File: D3/basic.ts
var width = Math.max(960, window.innerWidth)
var height = Math.max(500, window.innerHeight)
var x1 = width / 2
var y1 = height / 2
var x0 = x1
var y0 = y1
var i = 0
var r = 200
var tau = Math.PI * 2

var canvas = d3.select("body").append("canvas")

canvas.attr("width", width).attr("height", height)

d3.select("canvas").style("background-color", "#222")

function move(data: any, index: number) {
    var mouse = d3.mouse(canvas.node())
    x1 = mouse[0]
    y1 = mouse[1]
    d3.event.preventDefault()
}

canvas.on("mousemove", move)

var element: any = canvas.node();
var elemCanvas = element;

var workbench = new visual.Workbench2D(elemCanvas, window)

var context = elemCanvas.getContext("2d")
context.globalCompositionOperation = "lighter"
context.lineWidth = 2

function setUp() {
    workbench.setUp()
}

function tick(t) {
    i += 1
    context.clearRect(0,0,width,height)
    var z = d3.hsl(i % 360, 1, 0.5).rgb()
    var c = "rgba(" + z.r + "," + z.g + "," + z.b + ","
    x0 += (x1 - x0) * .1
    y0 += (y1 - y0) * .1
    var x = x0
    var y = y0
    function tweeny(unused1, unused2) {
        function circle(t) {
            var s = c + "" + (1-t) + ")"
            context.strokeStyle = s
            context.beginPath()
            context.arc(x, y, r * t, 0, tau)
            context.stroke()
        }
        return circle
    }
    // The following is correct. Maybe an error in the TypeLibrary definitions?
    d3.select({}).transition().duration(2000).ease(Math.sqrt).tween("circle", tweeny)
}

function terminate(t) {
    return t > 60
}
    
function tearDown(e) {
    workbench.tearDown()
}

var war = eight.animationRunner(tick, terminate, setUp, tearDown, window)
war.start()
